src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -492,8 +501,17 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             _handleControlC();
         }
 
-        const auto lock = _terminal->LockForWriting();
-        return _terminal->SendCharEvent(ch, scanCode, modifiers);
+        TerminalInput::OutputType out;
+        {
+            const auto lock = _terminal->LockForReading();

```

Comment:
Reading? `SendCharEvent` manipulates state.

Reply:
Wait, SendKeyEvent manipulates state and SendCharEvent only reads it?

Reply:
They both manipulate state and I should've used `LockForWriting` here. I think this is fine though, as this is almost assuredly not the only place where this is incorrect. I've been considering to just replace it with `Lock()` or to finally wrap `Terminal` entirely into a `til::mutex<Terminal>` kind of struct (but with a fair mutex).
Should I do that?

Reply:
I wouldn't necessarily do either of those _for the backport version_ of this PR. I can see the ways in which the `til::mutex`-like wrapper would be great, though... what do you feel is the right answer?

Reply:
I think the wrapped mutex class has significant safety benefits, because you cannot forget to lock something. But it wouldn't have prevented the bug this PR addresses. I'm fairly confident that I've fixed all cases where we forgot to lock `Terminal` right now so I don't think it would have any additional benefit for a backport either.